### PR TITLE
[AArch64][NEON][SVE] Lower mixed sign/zero extended partial reductions to usdot

### PR DESCRIPTION
**Pull request by SamTebbs33**
This PR adds lowering for partial reductions of a mix of sign/zero extended inputs to the usdot intrinsic.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -21824,37 +21824,59 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
...
+      !(ReducedType == MVT::v4i32 && MulSrcType == MVT::v16i8) &&
+      !(ReducedType == MVT::v2i32 && MulSrcType == MVT::v8i8))
+    return SDValue();
 
-  return SDValue();
+  bool AIsSExt = ExtA->getOpcode() == ISD::SIGN_EXTEND;
+  bool AIsZExt = ExtA->getOpcode() == ISD::ZERO_EXTEND;
+  bool BIsSExt = ExtB->getOpcode() == ISD::SIGN_EXTEND;
+  bool BIsZExt = ExtB->getOpcode() == ISD::ZERO_EXTEND;
+  if (!(AIsSExt || AIsZExt) || !(BIsSExt || BIsZExt))
```
Comment (NickGuy-Arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21847]: Not sure if this check is necessary (though nor was it needed before this patch). The code that emits this intrinsic checks this condition (see https://github.com/llvm/llvm-project/pull/92418/files#diff-da321d454a7246f8ae276bf1db2782bf26b5210b8133cb59e4d7fd45d0905decR2156-R2158), so outside of hand-written IR the case of no extends is never taken.
That, and there doesn't seem to be any tests that check this condition either.

That said, I'm fine with this staying as a bit of defensive coding, as I can't say for sure whether all partial reduction cases in the future will match on extends, but figured it was worth bringing forward

**Review by NickGuy-Arm — APPROVED** [review]
Looks good to me

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 1 above)
Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21847]: This function must stand on it's own and thus cannot assume anything about its input that is not part of the intrinsic's definition.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -21824,37 +21824,59 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
 
   auto ExtA = MulOp->getOperand(0);
   auto ExtB = MulOp->getOperand(1);
-  bool IsSExt = ExtA->getOpcode() == ISD::SIGN_EXTEND;
-  bool IsZExt = ExtA->getOpcode() == ISD::ZERO_EXTEND;
-  if (ExtA->getOpcode() != ExtB->getOpcode() || (!IsSExt && !IsZExt))
-    return SDValue();
-
   auto A = ExtA->getOperand(0);
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21827]: You cannot just move these checks because you need to prove `ExtA` and `ExtB` have an operand before calling `getOperand()`.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -21824,37 +21824,59 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
...
+      return SDValue();
+    bool Scalable = N->getValueType(0).isScalableVT();
+
+    // There's no nxv2i64 version of usdot
+    if (Scalable && ReducedType != MVT::nxv4i32)
+      return SDValue();
+
+    unsigned IntrinsicID =
+        Scalable ? Intrinsic::aarch64_sve_usdot : Intrinsic::aarch64_neon_usdot;
+    // USDOT expects the first operand to be unsigned, so swap the operands if
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21862]: Given there is sufficient reason for `AArch64ISD::SDOT` and `AArch64ISD::UDOT` to exist I propose the same likely holds for creating `AArch64ISD::USDOT`? Which is perhaps best done under a separate PR.

**Review by paulwalker-arm — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21827]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 3 above)
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Yeah I agree this can be done separately.

Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: For the avoidance of doubt I am suggesting it's a requirement of this PR that `AArch64ISD::USDOT` exist rather than suggesting some post PR refactoring.

Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: I've created a USDOT node, please let me know if it's not as expected.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 4]
```diff
@@ -1420,6 +1421,9 @@ def USMMLA : SIMDThreeSameVectorMatMul<1, 0, "usmmla", int_aarch64_neon_usmmla>;
 defm USDOT : SIMDThreeSameVectorDot<0, 1, "usdot", int_aarch64_neon_usdot>;
 defm USDOTlane : SIMDThreeSameVectorDotIndex<0, 1, 0b10, "usdot", int_aarch64_neon_usdot>;
 
+def : Pat<(v4i32 (AArch64usdot (v4i32 V128:$Rd), (v16i8 V128:$Rm), (v16i8 V128:$Rn))), (USDOTv16i8 $Rd, $Rm, $Rn)>;
+def : Pat<(v2i32 (AArch64usdot (v2i32 V64:$Rd), (v8i8 V64:$Rm), (v8i8 V64:$Rn))), (USDOTv8i8 $Rd, $Rm, $Rn)>;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:1425]: You don't need extra patterns here but instead can pass `AArch64usdot` directly to `defm USDOT...` in place of the existing `int_aarch64_neon_usdot` parameter.

If you look at the way UDOT is handled you see the missing piece is a small update to `AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN` to lower the intrinsics to `AArch64ISD::USDOT`.  Doing this will mean all future optimisations will apply equally to all places where the operation exists.

File: llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td [hunk 5]
```diff
@@ -3408,6 +3408,7 @@ let Predicates = [HasSVEorSME, HasMatMulInt8] in {
   defm USDOT_ZZZ  : sve_int_dot_mixed<"usdot", int_aarch64_sve_usdot>;
   defm USDOT_ZZZI : sve_int_dot_mixed_indexed<0, "usdot", int_aarch64_sve_usdot_lane>;
   defm SUDOT_ZZZI : sve_int_dot_mixed_indexed<1, "sudot", int_aarch64_sve_sudot_lane>;
+  def : Pat<(nxv4i32 (AArch64usdot (nxv4i32 ZPR32:$Rd), (nxv16i8 ZPR8:$Rm), (nxv16i8 ZPR8:$Rn))), (USDOT_ZZZ $Rd, $Rm, $Rn)>;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3411]: As above, you can pass `AArch64usdot` directly into `defm USDOT_ZZZ :` replacing the existing `int_aarch64_sve_usdot` parameter.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 4 above)
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Thank you, that is cleaner. Done.

File: llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td (hunk 5 above)
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 6]
```diff
@@ -21824,37 +21830,54 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
...
   auto ExtB = MulOp->getOperand(1);
-  bool IsSExt = ExtA->getOpcode() == ISD::SIGN_EXTEND;
-  bool IsZExt = ExtA->getOpcode() == ISD::ZERO_EXTEND;
-  if (ExtA->getOpcode() != ExtB->getOpcode() || (!IsSExt && !IsZExt))
+
+  bool AIsSExt = ExtA->getOpcode() == ISD::SIGN_EXTEND;
+  bool AIsZExt = ExtA->getOpcode() == ISD::ZERO_EXTEND;
+  bool BIsSExt = ExtB->getOpcode() == ISD::SIGN_EXTEND;
+  bool BIsZExt = ExtB->getOpcode() == ISD::ZERO_EXTEND;
+  if (!(AIsSExt || AIsZExt) || !(BIsSExt || BIsZExt))
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21838]: Now we care about all combinations I think you'd be better of with something like:
```
if (!isExtOpcode(ExtA->getOpcode() || !isExtOpcode(ExtB->getOpcode())
  return SDValue();

bool AIsSigned = ExtA->getOpcode() == ISD::SIGN_EXTEND;
bool BIsSigned = ExtB->getOpcode() == ISD::SIGN_EXTEND;
```
Whilst this allows ISD::ANY_EXTEND to pass through, I believe they would make the result undefined and thus emitting any DOT still a valid option.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 7]
```diff
@@ -1417,7 +1418,7 @@ let Predicates = [HasMatMulInt8] in {
 def  SMMLA : SIMDThreeSameVectorMatMul<0, 0, "smmla", int_aarch64_neon_smmla>;
 def  UMMLA : SIMDThreeSameVectorMatMul<0, 1, "ummla", int_aarch64_neon_ummla>;
 def USMMLA : SIMDThreeSameVectorMatMul<1, 0, "usmmla", int_aarch64_neon_usmmla>;
-defm USDOT : SIMDThreeSameVectorDot<0, 1, "usdot", int_aarch64_neon_usdot>;
+defm USDOT : SIMDThreeSameVectorDot<0, 1, "usdot", AArch64usdot>;
 defm USDOTlane : SIMDThreeSameVectorDotIndex<0, 1, 0b10, "usdot", int_aarch64_neon_usdot>;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:1422]: You'll need to replace all uses of `int_aarch64_neon_usdot` otherwise their instruction match will fail because the intrinsic no longer gets this far.  I'm somewhat surprised you did not see any test failures?

File: llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll [hunk 8]
```diff
@@ -1,6 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
-; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT
-; RUN: llc -mtriple aarch64 -mattr=+neon < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NODOT
+; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod,+i8mm < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT
+; RUN: llc -mtriple aarch64 -mattr=+neon,+i8mm < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NODOT
+; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NOIMM8
```
Comment (paulwalker-arm) [inline llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:4]: Would the following reduce the number of CHECK lines?
```
RUN_1: CHECK,CHECK-DOT,CHECK-I8MM
RUN_2: CHECK,CHECK-NODOT,CHECK-I8MM
RUN_3: CHECK,CHECK-DOT,CHECK-NOI8MM
```

**Review by paulwalker-arm — COMMENTED** [review]
A few suggestions, but only the potential bug in `AArch64InstrInfo.td` is holding me back from accepting the PR.

File: llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll [hunk 9]
```diff
@@ -1,6 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
-; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT
-; RUN: llc -mtriple aarch64 -mattr=+neon < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NODOT
+; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod,+i8mm < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT,CHECK-I8MM
+; RUN: llc -mtriple aarch64 -mattr=+neon,+i8mm < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NODOT,CHECK-I8MM
+; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT,CHECK-NOI8MM
```
Comment (paulwalker-arm) [inline llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:4]: This hasn't worked as expected because there are no `CHECK-I8MM` lines. I guess I was expecting too much from the auto-update script?

Comment (paulwalker-arm) [reply llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:4]: I think I see what has happened.  I had assumed `+i8mm` would enable `usdot` instructions but the DAG combine requires both `+dotprod` and `+i8mm` for that to happen.

Comment (SamTebbs33) [reply llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll]: That's nice. I re-ordered the run lines a little to reduce the number of total changes.

**Review by paulwalker-arm — APPROVED** [review]
